sled-agent/src/services.rs

```diff
@@ -203,12 +233,227 @@ impl ServiceManager {
+        boot_ctx.expect().return_once(|name| {
+            assert_eq!(name, EXPECTED_ZONE_NAME);
+            Ok(())
+        });
+        // Wait for the networking service.
+        let wait_ctx = svc::wait_for_service_context();
+        wait_ctx.expect().return_once(|_, _| Ok(()));
+        // Import the manifest, enable the service
+        let execute_ctx = crate::illumos::execute_context();
+        execute_ctx.expect().times(2).returning(|_| {
+            Ok(std::process::Output {
+                status: std::process::ExitStatus::from_raw(0),
+                stdout: vec![],
+                stderr: vec![],
+            })
+        });
+
+        vec![
+            Box::new(create_vnic_ctx),
+            Box::new(install_ctx),
+            Box::new(boot_ctx),
+            Box::new(wait_ctx),
+            Box::new(execute_ctx),
+        ]
+    }
+
+    // Prepare to call "ensure" for a new service, then actually call "ensure".
+    async fn ensure_new_service(mgr: &ServiceManager) {
+        let _expectations = expect_new_service();
+
+        mgr.ensure(ServiceEnsureBody {
+            services: vec![ServiceRequest {
+                name: SVC_NAME.to_string(),
+                addresses: vec![],
+            }],
+        })
+        .await
+        .unwrap();
+    }
+
+    // Prepare to call "ensure" for a service which already exists. We should
+    // return the service without actually installing a new zone.
+    async fn ensure_existing_service(mgr: &ServiceManager) {
+        mgr.ensure(ServiceEnsureBody {
+            services: vec![ServiceRequest {
+                name: SVC_NAME.to_string(),
+                addresses: vec![],
+            }],
+        })
+        .await
+        .unwrap();
+    }
+
+    // Prepare to drop the service manager.
+    //
+    // This will shut down all allocated zones, and delete their
+    // associated VNICs.
+    fn drop_service_manager(mgr: ServiceManager) {
+        let halt_ctx = MockZones::halt_and_remove_context();
+        halt_ctx.expect().returning(|_, name| {
+            assert_eq!(name, EXPECTED_ZONE_NAME);
+            Ok(())
+        });
+        let delete_vnic_ctx = MockDladm::delete_vnic_context();
+        delete_vnic_ctx.expect().returning(|_| Ok(()));
+
+        // Explicitly drop the servie manager
+        drop(mgr);
+    }
+
+    #[tokio::test]
+    #[serial_test::serial]
+    async fn test_ensure_service() {
+        let logctx =
+            omicron_test_utils::dev::test_setup_log("test_ensure_service");
+        let log = logctx.log.clone();
+
+        let config_dir = tempfile::TempDir::new().unwrap();
+        let config = config_dir.path().join("services.toml");
+        let mgr = ServiceManager::new(
+            log,
+            Some(PhysicalLink(EXPECTED_LINK_NAME.to_string())),
+            Some(config),
+        )
+        .await
+        .unwrap();
+
+        ensure_new_service(&mgr).await;
+        drop_service_manager(mgr);
```

Comment:
I think you need the code to indicate the test was successful: `logctx.cleanup_successful()`.

Reply:
ugh, I always forget this! Thanks, added.